rocmlibs: cleanup amdgpu_target variant behaviour

cgmb opened the pull request:
This change builds on #28846 to apply a consistent amdgpu_target variant behaviour across all the ROCm math and communication libraries. There's two main things I've done:

1. Define the amdgpu_target variant values with `auto_or_any_combination_of` on all ROCm math and communication libraries that contain GPU code. When `auto` is supplied, the `AMDGPU_TARGETS` cache variable is left unset. In this case, the libraries will default to building for all supported architectures. This takes a _long_ time, but it's the safest default.
2. Propagate the amdgpu_target variant between ROCm math library dependencies. e.g. `spack spec rocsolver amdgpu_target=gfx906:xnack-,gfx1010` will include `^rocblas@4.5.2%gcc@9.3.0~ipo amdgpu_target=gfx1010,gfx906:xnack-`

This change was motivated by comments in https://github.com/spack/spack/pull/28846#discussion_r805200534 and https://github.com/spack/spack/pull/27535#issuecomment-987482569.

teonnik committed:
rocBLAS: Add an option to disable tensile

- Change variant name to be consistent with rocsolver
- Change default variant value to be consistent with rocsolver

cgmb committed:
Make consistent rocmlibs amdgpu_target variants

cgmb committed:
Propagate amdgpu_target variant

cgmb committed:
Fix Python 2 compatibility

teonnik commented on var/spack/repos/builtin/packages/rocblas/package.py:
```diff
@@ -29,27 +29,28 @@ class Rocblas(CMakePackage):
 
-    variant('tensile_architecture', default='all', values=tensile_architecture, multi=True)
+    variant('amdgpu_target', values=auto_or_any_combination_of(*amdgpu_targets))
+    variant('tensile', default=False, description='Use Tensile as a backend')
```
LGTM! The only change needed is to make `default=True` otherwise `gemm` symbols will be missing as discussed in the #28846 .

cgmb replied:
I will rebase once your PR is merged.

teonnik committed:
Set default for tensile to True to avoid surprises with missing symbols

teonnik committed:
Use auto_or_any_combination_of for AMD GPU architecture variant

cgmb commented:
I rebased #28846 on develop (resolving the conflicts) and then rebased this on that. If we want to merge the PRs seperately, we could hard-reset #28846 to a1c4cf088b138efe56e56a888145ecb03281ff68 and merge that, or just close it and merge this instead.

teonnik commented:
I closed #28846, we can just have this one merged.

cgmb committed:
Fix amdgpu_target=auto

Since amdgpu_target is a MultiValuedVariant, the value is always
actually a tuple. However, all variants define __contains__ to
handle these sorts of comparisons.

cgmb committed:
Fix amdgpu_target=auto setting HCC_AMDGPU_TARGET

cgmb committed:
Tighten HIP CMake requirements

cgmb committed:
Fix spack style

cgmb commented:
@spackbot run pipeline

cgmb commented:
@spackbot run pipeline

cgmb commented:
@haampie, @srekolam, @arjun-raj-kuppala

This is ready for review. I've tested all the changed libraries with ROCm 4.5 and 5.0 and tried a variety of configurations. I will continue testing with older versions, and update if I find any problems.

With this change, the ROCm math and communication libraries in Spack will be built by default with the same architectures used for the upstream AMD ROCm builds. However, you will be able to choose to build for other architectures _and_ those choices will be correctly propagated between dependencies.

I suspect that the difficulties correctly installing ROCm via Spack exhibited in [this thread](https://github.com/spack/spack/pull/27535#issuecomment-985031514) are common among novice users, and believe that these changes will greatly reduce user confusion.

cgmb commented:
I also tested a set of selection variants with ROCm 4.3.1 and ROCm 4.2.0.

cgmb commented:
Thanks @haampie!
